### DISCUSSION TITLE
How to Use Uploaded Data

### DISCUSSION THREAD

**Discussion by mtanco** [h2oai/wave#684] [Q&A]
Can you ever use data that is uploaded to the wave server directly, or does it always need to be downloaded to the Python environment to be used?

If this download step is not necessary, could we get an example of uploaded data being used in some component directly?

Today: 
```
q.client.working_file_path = await q.site.download(url=q.args.file_upload[0], path=data_path)
# read data from q.client.working_file_path and do put in some ui component
```

Question:
```
# q.args.file_upload[0] goes directly to some component
```

**Comment by lo5** [comment]
This is currently not possible. The download step is required.

**Comment by ryanchesler** [comment]
Having a bit of a hard time with something related to this. Not sure exactly what the intended workflow is with this. 

It seems most powerful to have all of the files organized into users own folders within the app, but occasionally files need to be uploaded to the server to be displayed in the browser, is that correct or can you show images within the browser directly from the app?

Trying to simplify the image annotation interface. VIA is able to import the data when it is in the wave server data folder, but not when it is on the app even when I give it absolute paths. VIA then exports the uploaded wave server filepaths which makes it a bit cumbersome because then I need to maintain a reverse lookup and be manually rewriting a lot of paths in the VIA annotation file. If I never unload the images then things work, but it accumulates a lot of files on the server, many redundant. If I unload the files then it is impossible to review the old annotations because the same files will be uploaded to a new subdirectory. 

Not sure if there is any way around this. Seems like the best solution right now is just to accumulate a ton of files on the server and have duplicates on the app, but that seems sub-optimal long term.

**Reply by lo5** [reply]
The primary reason files need to be uploaded to the server is that the app and the server could be running on different machines without access to the same filesystem. The plan is to [automatically delete unused files from the server](https://github.com/h2oai/wave/issues/686). 

> Seems like the best solution right now is just to accumulate a ton of files on the server and have duplicates on the app, but that seems sub-optimal long term.

That is what I would do for now - have the app be the single source of truth, and not worry about the duplicates accumulating on the server-side.

**Comment by AakashKumarNain** [comment]
`Trying to simplify the image annotation interface.`

Totally. For any image based workflow, these are some of the common things that an end user would want to do within a wave app:
1. List all the folders containing images
2. Browse through the images to check for things like image quality, anomalies, corrupt images, etc
3. Annotation can be done in a much easier way if 2) is possible
